QUICK-FIX Update custom attribute API

zidarsk8 opened the pull request:
This PR refactors the custom attributable mixin so that the api looks cleaner.

We no longer have to send  two separate arrays with definitions and values to store custom attribute values and make extra get requests to receive these new values. For compatibility the old code will remain untill the frontend changes are made.

~~This PR is still work in progress, but should be a good indicator to see where this is going. Feel free to comment on any design or style issues that you can find.~~

Note: When testing please click through all parts of the app that touch custom attributes.

alieh-rymasheuski commented on src/ggrc/builder/json.py:
```diff
@@ -78,7 +81,7 @@ def publish_stub(obj, inclusions=(), inclusion_filter=None):
     ret['type'] = obj.__class__.__name__
     ret['context_id'] = obj.context_id
-    if hasattr(publisher, '_stub_attrs') and publisher._stub_attrs:
+    if getattr(publisher, '_stub_attrs', False):
```
This should be `getattr(publisher, '_stub_attrs', [])`, because `_stub_attrs` is a `list` if it is present, so that this code doesn't confuse the reader into thinking `_stub_attrs` is `bool`.

zidarsk8 replied:
Updated, thanks.

alieh-rymasheuski commented on src/ggrc/models/custom_attribute_value.py:
```diff
@@ -160,3 +168,42 @@ def _extra_table_args(_):
+    is_mandatory = self.custom_attribute.multi_choice_mandatory
+    if self.attribute_value and self.attribute_value not in valid_options:
+      raise ValueError("Invalid custom attribute dropdown option")
+    elif not self.attribute_value and is_mandatory:
```
I'm not sure you use `multi_choice_mandatory` correctly. This field stores bitmaps for mandatory comments and evidences and has nothing to do with `attribute_value` being mandatory.

alieh-rymasheuski replied:
At the moment it is possible to save some value and _then_ `None` into a mandatory Dropdown CA via inline edit. I'm not sure we should forbid saving an empty mandatory CA, as the validation for mandatory value presence should be done only on completion of an Assessment.

zidarsk8 replied:
good catch thanks

alieh-rymasheuski commented on src/ggrc/builder/json.py:
```diff
@@ -128,20 +132,23 @@ def do_update_attr(cls, obj, json_obj, attr):
       # CustomAttributable mixin
       attr_name = attr
       value = class_attr(obj, json_obj)
+    elif update_raw:
```
Although the code is quite clear (what can't be said about the code under `else` clause :)), I'd expect a comment here to make every conditional branch look alike.

zidarsk8 replied:
Added, and please check if it's descriptive enough.

alieh-rymasheuski replied:
Quite good, thanks. :)

Smotko commented on src/ggrc/services/common.py:
```diff
@@ -1234,6 +1237,12 @@ def collection_post_loop(self, body, res, no_result, running_async):
     with benchmark("Send collection POSTed event"):
       self.collection_posted.send(obj.__class__, objects=objects)
+    with benchmark("Flush posted objects"):
+      db.session.flush()
```
I believe the tests are failing because of this flush. Maybe we need to move it down after the permission check?

zidarsk8 commented:
I am adding Please review label here for two reasons:
- quite a few bugs are related to this PR, and this merge is needed for a proper solution.
- This is quite a big change of how our API handles custom attributes and it needs to be thoroughly tested, preferably by multiple reviewers.

Smotko commented on test/integration/ggrc/models/mixins/test_customattributable.py:
```diff
@@ -0,0 +1,205 @@
+    prog = factories.ProgramFactory()
+    cad = factories.CustomAttributeDefinitionFactory
+    cad1 = cad(definition_type="program", title="CA 1",)
+    cad3 = cad(definition_type="program", title="CA 3",)
```
The names for these feel a bit off. What happened to `cad2` and I think it's a bit confusing that the CustomAttributeDefinitionFactory class is aliased as `cad`, I think you should name it either `Cad` or `CAD` to differentiate between instances...

Smotko commented on test/integration/ggrc/models/mixins/test_customattributable.py:
```diff
@@ -0,0 +1,205 @@
+
+  def test_updating_ca_values(self):
+    """Test updating custom attribute values."""
+    cad = factories.CustomAttributeDefinitionFactory
```
Also, if you are doing this in every single test it's probably better to just alias it on import. I know this is against google's import guidelines, but this workaround is probably even worse :)

zidarsk8 replied:
It's actually not against google's guidelines as I see it. So yeah, I'll update this.

Smotko commented on src/ggrc/models/mixins/customattributable.py:
```diff
@@ -86,6 +87,72 @@ def join_function():
+      else:
+        new_value.attributable = self
+        self._custom_attribute_values.append(new_value)
+    pass
```
Pass here is redundant.

zidarsk8 replied:
Leftovers :)

Smotko commented:
I received the following error when updating custom attributes on a facility:

```
Traceback (most recent call last):
  File "/vagrant/src/ggrc/services/common.py", line 732, in dispatch_request
    return self.put(*args, **kwargs)
  File "/vagrant/src/ggrc/services/common.py", line 842, in put
    self.json_update(obj, src)
  File "/vagrant/src/ggrc/services/common.py", line 807, in json_update
    ggrc.builder.json.update(obj, src)
  File "/vagrant/src/ggrc/builder/json.py", line 98, in update
    updater.update(obj, json_obj)
  File "/vagrant/src/ggrc/builder/json.py", line 751, in update
    self.update_attrs(obj, json_obj)
  File "/vagrant/src/ggrc/builder/json.py", line 724, in update_attrs
    self.do_update_attrs(obj, json_obj, self._update_attrs)
  File "/vagrant/src/ggrc/builder/json.py", line 718, in do_update_attrs
    UpdateAttrHandler.do_update_attr(obj, json_obj, attr_name)
  File "/vagrant/src/ggrc/builder/json.py", line 182, in do_update_attr
    setattr(obj, attr_name, value)
  File "/vagrant-dev/opt/dev_virtualenv/local/lib/python2.7/site-packages/sqlalchemy/ext/hybrid.py", line 747, in __set__
    self.fset(instance, value)
  File "/vagrant/src/ggrc/models/mixins/customattributable.py", line 122, in custom_attribute_values
    self._add_ca_value_dicts(values)
  File "/vagrant/src/ggrc/models/mixins/customattributable.py", line 170, in _add_ca_value_dicts
    raise BadRequest("Bad custom attribute value inserted")
BadRequest: 400: Bad Request
```

PUT JSON:
```
{"facility":{"custom_attribute_values":[{"id":45657,"href":"/api/custom_attribute_values/45657","type":"CustomAttributeValue"},{"id":45658,"href":"/api/custom_attribute_values/45658","type":"CustomAttributeValue"},{"id":45659,"href":"/api/custom_attribute_values/45659","type":"CustomAttributeValue"},{"id":45660,"href":"/api/custom_attribute_values/45660","type":"CustomAttributeValue"}],"people":[],"updated_at":"2016-08-05T11:11:03Z","task_group_objects":[],"secondary_contact":null,"id":2317,"description":"","documents":[],"modified_by":{"id":323,"href":"/api/people/323","type":"Person"},"title":"asdf asdf","contact":{"id":323,"href":"/api/people/323","type":"Person"},"object_people":[],"type":"Facility","start_date":"","viewLink":"/facilities/2317","status":"Draft","related_sources":[],"end_date":"","workflow_state":null,"os_state":"Modified","risk_objects":[],"object_documents":[],"related_destinations":[],"slug":"FACILITY-2317","owners":[{"id":323,"href":"/api/people/323","type":"Person"}],"url":"","task_groups":[],"notes":"","reference_url":"","object_owners":[{"context_id":null,"href":"/api/object_owners/35684","type":"ObjectOwner","id":35684}],"custom_attribute_definitions":[{"mandatory":false,"attribute_type":"Dropdown","updated_at":"2015-12-01T20:57:32","definition_type":"facility","helptext":"Business Unit","placeholder":null,"id":168,"multi_choice_options":"Legal,HR,Finance,InfoSec,Engineering,IT","modified_by":{"context_id":null,"href":"/api/people/152","type":"Person","id":152},"title":"Business Unit","multi_choice_mandatory":null,"created_at":"2015-12-01T19:54:19","definition_id":null,"context":null,"type":"CustomAttributeDefinition","selfLink":"/api/custom_attribute_definitions/168"},{"mandatory":false,"attribute_type":"Dropdown","updated_at":"2015-12-01T19:54:51","definition_type":"facility","helptext":"Type of Facility","placeholder":null,"id":169,"multi_choice_options":"Data Center,HQ,Satellite,Contract","modified_by":{"context_id":null,"href":"/api/people/152","type":"Person","id":152},"title":"Type of Facility","multi_choice_mandatory":null,"created_at":"2015-12-01T19:54:51","definition_id":null,"context":null,"type":"CustomAttributeDefinition","selfLink":"/api/custom_attribute_definitions/169"},{"mandatory":false,"attribute_type":"Date","updated_at":"2015-12-01T19:55:11","definition_type":"facility","helptext":"Launch date","placeholder":null,"id":170,"multi_choice_options":null,"modified_by":{"context_id":null,"href":"/api/people/152","type":"Person","id":152},"title":"Launch date","multi_choice_mandatory":null,"created_at":"2015-12-01T19:55:11","definition_id":null,"context":null,"type":"CustomAttributeDefinition","selfLink":"/api/custom_attribute_definitions/170"},{"mandatory":false,"attribute_type":"Rich Text","updated_at":"2016-08-05T11:10:29","definition_type":"facility","helptext":"","placeholder":"","id":186,"multi_choice_options":null,"modified_by":{"context_id":null,"href":"/api/people/323","type":"Person","id":323},"title":"asdf","multi_choice_mandatory":null,"created_at":"2016-08-05T11:10:29","definition_id":null,"context":null,"type":"CustomAttributeDefinition","selfLink":"/api/custom_attribute_definitions/186"}],"risks":[],"context":null,"created_at":"2016-08-05T11:11:03Z","selfLink":"/api/facilities/2317","_pending_joins":[],"custom_attributes":{"168":"InfoSec","169":"Contract","170":"08/18/2016","186":"asdfsadfa dfsf sdaf"},"_transient":{},"_transient:title":null,"href":"/api/facilities/2317"}}
```

zidarsk8 commented:
@Smotko 💯 kudos for the nice Bug report.

I have updated the tests to catch this issue and fixed the check for stubs that are sent with existing custom attribute values.

Smotko commented:
PUTs still seem to fail, but in a different way as before :/

```
Traceback (most recent call last):
  File "/vagrant-dev/opt/dev_virtualenv/lib/python2.7/site-packages/flask/app.py", line 1836, in __call__
    return self.wsgi_app(environ, start_response)
  File "/vagrant-dev/opt/dev_virtualenv/lib/python2.7/site-packages/flask/app.py", line 1820, in wsgi_app
    response = self.make_response(self.handle_exception(e))
  File "/vagrant-dev/opt/dev_virtualenv/lib/python2.7/site-packages/flask/app.py", line 1403, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/vagrant-dev/opt/dev_virtualenv/lib/python2.7/site-packages/flask/app.py", line 1817, in wsgi_app
    response = self.full_dispatch_request()
  File "/vagrant-dev/opt/dev_virtualenv/lib/python2.7/site-packages/flask/app.py", line 1477, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/vagrant-dev/opt/dev_virtualenv/lib/python2.7/site-packages/flask/app.py", line 1381, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/vagrant-dev/opt/dev_virtualenv/lib/python2.7/site-packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/vagrant-dev/opt/dev_virtualenv/lib/python2.7/site-packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/vagrant-dev/opt/dev_virtualenv/lib/python2.7/site-packages/flask_login.py", line 637, in decorated_view
    return func(*args, **kwargs)
  File "/vagrant-dev/opt/dev_virtualenv/lib/python2.7/site-packages/flask/views.py", line 84, in view
    return self.dispatch_request(*args, **kwargs)
  File "/vagrant/src/ggrc/services/common.py", line 732, in dispatch_request
    return self.put(*args, **kwargs)
  File "/vagrant/src/ggrc/services/common.py", line 849, in put
    obj.validate_custom_attributes()
  File "/vagrant/src/ggrc/models/mixins/customattributable.py", line 388, in validate_custom_attributes
    value.validate()
  File "/vagrant/src/ggrc/models/custom_attribute_value.py", line 200, in validate
    raise ValueError("Custom attribute definition not found: Can not "
ValueError: Custom attribute definition not found: Can not validate custom attribute value
```

zidarsk8 commented:
This needs to be on hold until https://github.com/google/ggrc-core/pull/4196 gets merged, because there will be conflicts.

But that PR should have been split into one with backend part and one with frontend part.

Smotko commented:
I've checked the old API and as far as I can tell nothing is broken. I've also tried to use the new API with: 

```
facility.refresh().then(function(f) {delete f.custom_attributes; f.custom_attribute_values[0].attr('attribute_value', 'HQ'); f.save()});
```

and it saved the value without a problem. The only issue I was able to find seems to minor:

Validator is throwing a 500 (server error) on invalid data, but it should have been throwing a 400 (bad request). @zidarsk8 please let me know if this is something you want to fix in this PR or in a separate one.

zidarsk8 commented:
@Smotko The validator throws a `ValueError` which we're not handling properly. This is already fixed in https://github.com/google/ggrc-core/pull/4224 so if you test this with both PRs merged it should work fine.
